seqr/views/apis/anvil_workspace_api_tests.py

```diff
@@ -422,14 +436,47 @@ def test_validate_anvil_vcf(self, mock_file_iter, mock_file_exist, mock_utils_lo
             'Incorrect meta Type for FORMAT.DP - expected "Integer", got "String"'
         ])
 
-        # Test valid operation
-        mock_file_exist.return_value = True
-        mock_file_iter.return_value = BASIC_META + INFO_META + FORMAT_META + HEADER_LINE + DATA_LINES
+        # Test valid operations
+        mock_subprocess.reset_mock()
+        mock_file_logger.reset_mock()
+        mock_subprocess.return_value.stdout = BASIC_META + INFO_META + FORMAT_META + HEADER_LINE + DATA_LINES
         response = self.client.post(url, content_type='application/json', data=json.dumps(VALIDATE_VCF_BODY))
         self.assertEqual(response.status_code, 200)
         self.assertDictEqual(response.json(), VALIDATE_VFC_RESPONSE)
-        mock_file_exist.assert_called_with('gs://test_bucket/test_path.vcf', user=self.manager_user)
-        mock_file_iter.assert_called_with('gs://test_bucket/test_path.vcf', byte_range=None)
+        mock_subprocess.assert_has_calls([
+            mock.call('gsutil ls gs://test_bucket/test_path.vcf', stdout=-1, stderr=-2, shell=True),
+            mock.call().wait(),
+            mock.call('gsutil cat gs://test_bucket/test_path.vcf', stdout=-1, stderr=-2, shell=True),
+        ])
+        mock_file_logger.info.assert_has_calls([
+            mock.call('==> gsutil ls gs://test_bucket/test_path.vcf', self.manager_user),
+            mock.call('==> gsutil cat gs://test_bucket/test_path.vcf', None),
+        ])
+
+        # Test a valid sharded VCF file path
+        mock_subprocess.reset_mock()
+        mock_file_exist_or_list_subproc = mock.MagicMock()
+        mock_get_header_subproc = mock.MagicMock()
+        mock_subprocess.side_effect = [mock_file_exist_or_list_subproc] * 3 + [mock_get_header_subproc]
+        mock_file_exist_or_list_subproc.wait.return_value = 0
+        mock_get_header_subproc.wait.return_value = 0
+        mock_file_exist_or_list_subproc.stdout = [b'gs://test_bucket/test_path-001.vcf.gz', b'gs://test_bucket/test_path-102.vcf.gz']
+        mock_get_header_subproc.stdout = BASIC_META + INFO_META + FORMAT_META + HEADER_LINE + DATA_LINES
+        response = self.client.post(url, content_type='application/json', data=json.dumps(REQUEST_BODY_SHARDED_DATA_PATH))
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json(), {'fullDataPath': 'gs://test_bucket/test_path-*.vcf.gz', 'vcfSamples': ['HG00735', 'NA19675', 'NA19678']})
+        mock_subprocess.assert_has_calls([
+            mock.call('gsutil ls gs://test_bucket/test_path-*.vcf.gz', stdout=-1, stderr=-2, shell=True),
+            mock.call('gsutil ls gs://test_bucket/test_path-*.vcf.gz', stdout=-1, stderr=-2, shell=True),
+            mock.call('gsutil ls gs://test_bucket/test_path-*.vcf.gz', stdout=-1, stderr=-2, shell=True),
```

Comment:
why is `ls` called 3 times in a row?

Reply:
One time for file existence, one for avoiding the empty bucket exception, and the other for fetching the sharded file list.

Reply:
>  one for avoiding the empty bucket exception

The empty bucket exception occurs when you run `ls` with the `**` to check the nested directories, so for this were you are never checking the nested directories it should not be run.

Reply:
Okay, I see. I can move the empty bucket detection into the `ls` with the `**` branch.